core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java

```diff
@@ -538,107 +539,103 @@ public void write(String s, byte[] bytes) {
-                .setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+        MethodCreator registerForReflection = addMethodRegisterForReflection(file);
+        MethodCreator registerClass = addMethodRegisterClass(file);
 
+        int index = 0;
+        MethodCreator currentRegisterClass = null;
         for (Map.Entry<String, ReflectionInfo> entry : reflectiveClasses.entrySet()) {
-            MethodCreator mv = file.getMethodCreator("registerClass" + count++, void.class, Feature.BeforeAnalysisAccess.class);
-            mv.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
-            registerForReflection.invokeStaticMethod(mv.getMethodDescriptor(), registerForReflection.getMethodParam(0));
-
-            TryBlock tc = mv.tryBlock();
-
-            ResultHandle clazz = tc.loadClassFromTCCL(entry.getKey());
-            //we call these methods first, so if they are going to throw an exception it happens before anything has been registered
-            ResultHandle constructors = tc
-                    .invokeVirtualMethod(ofMethod(Class.class, "getDeclaredConstructors", Constructor[].class), clazz);
-            ResultHandle methods = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredMethods", Method[].class), clazz);
-            ResultHandle fields = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredFields", Field[].class), clazz);
-
+            // To avoid getting MethodTooLargeException, the methods only manage up to 100 entries
+            // To avoid getting ClassTooLargeException, the methods need to manage several (100) entries
+            if (index++ % 100 == 0) {
+                if (currentRegisterClass != null) {
+                    currentRegisterClass.returnVoid();
+                }
+                currentRegisterClass = file.getMethodCreator("registerClass" + count++, void.class,
+                        Feature.BeforeAnalysisAccess.class);
+                currentRegisterClass.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+            }
+            boolean tryBlock = !entry.getValue().weak && (!entry.getValue().constructors && !entry.getValue().ctorSet.isEmpty()
+                    || !entry.getValue().methods && !entry.getValue().methodSet.isEmpty()
+                    || !entry.getValue().fields && !entry.getValue().fieldSet.isEmpty()) || entry.getValue().serialization;
+            final BytecodeCreator creator;
+            ResultHandle clazz = null;
+            if (tryBlock) {
+                TryBlock tc = currentRegisterClass.tryBlock();
+                CatchBlockCreator cc = tc.addCatch(Throwable.class);
+                cc.invokeVirtualMethod(ofMethod(Throwable.class, "printStackTrace", void.class), cc.getCaughtException());
+                creator = tc;
+                clazz = creator.loadClassFromTCCL(entry.getKey());
+            } else {
+                creator = currentRegisterClass;
+            }
```

Comment:
> This seems unnecessarily complex to me. Could you please elaborate on why you need to conditionally create the try block, while previously it was always created? Is this to save space?

The idea is to avoid having a try/catch block like before for each class to register to indeed save space. It is only added when one of `methodSet`, `fieldSet` or `ctorSet` is not empty or `serialization` is true.

> 
> If so, why not create a single try-catch for each `registerClass*` method and wrap the registration of the 100 entries in it? This would both save more space and simplify things.

Yes but it would then be a behavior change, if one class is missing, many classes won't be registered. For example in my use case it tries to register `com.sun.xml.internal.bind.v2.ContextFactory` which is not part of GraalVM so it silently throws a `ClassNotFoundException` which has no consequence with try/catch for each class if I propose a global try/catch I will lost several classes

Reply:
Uh, right! Thanks for clearing this up.

In that case could you please make the `tryBlock` check a bit more intuitive to the reader?
Maybe breaking it to a few booleans like `hasMethodSet`, `hasFieldSet`, etc.